doc/man/detops.1
.\" mcaops
.\" Contact Eugen Wintersberger <eugen.wintersberger@desy.de> for typos and corrections
.TH man 1 "12.10.2013" "@PROGRAM_VERSION@" "detops man page"
.SH NAME
 detops - apply cumulative image operations on detector files
.SH SYNOPSIS
.B detops [OPTIONS] [FILE]

.SH DESCRIPTION
\fBdetops\fR is a utility program applying certain operations on 2D detector data. 
The detector data can be read from one or several single image detector files.
.sp
The operations have typically 
All results are written to standard output. This allows several \fBmcaops\fR
commands to be chained using the pipe operator. 
.sp
The behavior of \fBmcaops\fR is controlled via the command string which is
passed as the first argument to the program (see the SYNOPSIS above).  Currently
the following commands are available:
.sp
\fBmax\fR     find the value and its position in the spectrum
.sp
\fBsum\fR     add all value in the spectrum
.sp
\fBrebin\fR   rebin the spectrum to new bin sizes
.sp
\fBscale\fR   scale the bin centers of the spectrum
.sp
\fBdump\fR    dump channel and mca data unchanged to standard out
.P 
These operations can roughly devided into two groups: accumulating and
non-accumulating operations. The former one reduces the entire MCA data to a
single number while the latter one preserves the spectrum character of the data
(though rebining maybe alters the number of bins). 

.SH OPTIONS
The  options provided and their meaning depend on the operation that should be
performed on the data. Some of of the options are global and valid for all
operations other are only defined for a particular command. 
.sp
Global options:
.TP
\fB\-h\fR, \fB\-\-help\fR
Print a usage message and exit.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Produce verbose output, printing information regarding the specified options and
objects.  All output is printed to standard error allowing you to redirect
payload data to an other program while still watching debugging output.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Suppress all unnecessary output (the counterpart of -v)
.TP 
\fB\-\-header\fR
print a header before dumping the output to standard out
.TP
\fB\-\-xolumn=[COLNAME]\fR
specify the name of the column to choose from the input to use as bin-center or
-index values. This option is only useful when used in connection with FIO
files.
.TP
\fB\-\-ycolumn=[COLNAME]\fR
specify the name of the column to use for MCA data. This option is only useful
with a FIO file. 
.P
Options for the \fBrebin\fR operation:
.TP
\fB\-b\fR,\fB\-\-binsize=[NBINS]\fR
defines the number of bins that should be collated during rebining. 
.TP 
\fB\-\-noxrebin\fR 
do not rebin the x-axis. Instead use indices for the output. 
.TP
\fB\-\-normalize\fR 
normalize the rebinned data.
.P 
Options for the  \fBscale\fR operation:
.TP
\fB\-c\fR,\fB\-\-center=[CENTER]\fR
defines the index of the center bin used for rescaling. 
.TP
\fB\-d\fR,\fB\-\-delta=[DELTA]\fR
defines the step width for the rebin operation
.TP 
\fB\-x\fR,\fB\-\-cvalue=[CENTERVALUE]\fR
the value of the center bin.
.P 
The \fBsum\fR and \fBmax\fR command have no additional options.

.SH EXIT STATUS
\fBmcaops\fR returns 0 in case of success and -1 otherwise. 

.SH EXAMPLES
Lets start with a very simple example reading data form standard input and using
the accumulating operations \fBsum\fR and \fBmax\fR. For the sake of convenience 
a data file should be created which looks like this
.IP
0 1
.IP
1 0
.IP
2 10
3 50
4 100
5 200
6 70
7 150
8 20
9 1
10 0
.P
and store the file as test.dat. Here the first column is the channel index while
the second holds the channel data. 
To compute the sum of the data stored use
.TP 
$> cat test.dat | mcaops sum 
602
.P 
where the output, 602, is the sum of values stored in the second column.
Analogously, to find the maximum use
.TP 
$> cat test.dat | mcaops max
5   200
.P 
the two  numbers are the position of the maximum and its value respectively. 
.SH AUTHOR
Written by Eugen Wintersberger.

.SH SEE ALSO



RELEASE_NOTES
Release notes for version 0.3.0
===============================

detinfo
-------
  * program output has been changed
  * several output formats including CSV and XML have been added
  * added support for Nexus files (fixes issue 2)

nxls
----
  * by default only short output is shown (fixes issue 3)
  * added additional information with the long option (fixes issue 6)

det2nx
------
  * removed sanity check of input files - this operation is too expensive
  * det2nx cannot create the basic file structure anymore
    this should avoid incomplete Nexus files. Create the file where your 
    detector data should be dumped before for instance with xml2nx. 
    However, the field where the data shall be stored can still be 
    created by det2nx. See the info pages for a detailed description 
    of the new behavior of the program.


src/detops/detops.cpp
//
// (c) Copyright 2013 DESY, Eugen Wintersberger <eugen.wintersberger@desy.de>
//
// This file is part of pnitools.
//
// pnitools is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// pnitools is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with pnitools.  If not, see <http://www.gnu.org/licenses/>.
// ===========================================================================
// Created on: Oct 28, 2013
//     Author: Eugen Wintersberger <eugen.wintersberger@desy.de>
///

#include <iomanip>
#include <fstream>
#include "detops.hpp"
#include "io_utils.hpp"
#include "roi_utils.hpp"
#include "image_processor.hpp"
#include "integrate.hpp"
#include "minimum.hpp"
#include "maximum.hpp"


void assemble_process_chain(const configuration &config,image_processor &proc)
{
    auto cmd = config.value<string>("command");

    if(cmd == "sum")
        proc.push_back(image_processor::op_ptr_type(new integrate()));
    else if(cmd == "max")
        proc.push_back(image_processor::op_ptr_type(new maximum()));
    else if(cmd == "min")
        proc.push_back(image_processor::op_ptr_type(new minimum()));
}

int main(int argc, char **argv) 
{

    configuration config = create_config();

    if(parse_cli_opts(argc,argv,"detops",config)) return 1;

    if(check_help_request(config,"")) return 1;

    //=========================basic input reading=============================
    //we try to do as much as possible before reading the list of input files.
    //As this list can be long and thus checking the files might take a while we
    //do not want to recognize right after this expensive job that we hat to
    //throw an exception due to some simple operation

    //select the command
    image_processor proc;
    
    //===============read darkfield and flatfield images========================
    try
    {
        //read dark- and flatfield image
        /*
        image_type dark_image = get_darkfield(config);
        image_type flat_image = get_flatfield(config);
        */

        //setup rois (if there are any)
        roi_stack rois = get_roi_stack(config);
   
        //generate the list of input files
        file_list input_files = get_input_files(config);

        //now we can start with the construction of the operation
        assemble_process_chain(config,proc);
     
#ifdef NOFOREACH
        BOOST_FOREACH(auto f,input_files)
#else
        for(auto f: input_files)
#endif
        {
           //read image from file
           image_type image = read_image(f); 

           /*
           std::ofstream of("test.dat");
           for(auto v: image) of<<v<<std::endl;
           of.close();*/
           //generate image stack from the image and rois
           image_stack istack;
           if(rois.empty())
               istack.push_back(image);
           else
               //here we most probably have to pass the dark and flat field
               //images - that would be the simplest approach
               istack= image_stack_from_rois(image,rois);

           //apply operation on each image
#ifdef NOFOREACH
           BOOST_FOREACH(auto i,istack)
#else
           for(auto i: istack) 
#endif
               proc(i);

           //show the result for the last operation
           std::cout<<std::setprecision(16)<<std::scientific;
           std::cout<<*proc.back()<<std::endl;
        }
    }
    catch(cli_option_error &error)
    {
        std::cerr<<error<<std::endl; return 1;
    }
    catch(file_error &error)
    {
        std::cerr<<error<<std::endl; return 1;
    }
    catch(parser_error &error)
    {
        std::cerr<<error<<std::endl; return 1;
    }
    /*
    catch(...)
    {
        std::cerr<<"Something went wrong!"<<std::endl;
        return 1;
    }*/

	return 0;
}

